Mathlib/NumberTheory/LegendreSymbol/GaussSum.lean

```diff
@@ -0,0 +1,373 @@
+  cases' eq_or_ne b 0 with hb hb
+  · -- case `b = 0`
+    simp only [hb, inv_zero, MulZeroClass.mul_zero, MulChar.map_zero, MulZeroClass.zero_mul,
+      Finset.sum_const_zero, map_zero_one, mul_one]
+    exact (hχ.sum_eq_zero).symm
+  · -- case `b ≠ 0`
+    refine' (Fintype.sum_bijective _ (mulLeft_bijective₀ b hb) _ _ fun x => _).symm
+    rw [mul_assoc, mul_comm x, ← mul_assoc, mul_inv_cancel hb, one_mul, mul_sub, mul_one]
+
+/-- We have `gaussSum χ ψ * gaussSum χ⁻¹ ψ⁻¹ = Fintype.card R`
+when `χ` is nontrivial and `ψ` is primitive (and `R` is a field). -/
+theorem gaussSum_mul_gaussSum_eq_card {χ : MulChar R R'} (hχ : IsNontrivial χ) {ψ : AddChar R R'}
+    (hψ : IsPrimitive ψ) : gaussSum χ ψ * gaussSum χ⁻¹ ψ⁻¹ = Fintype.card R := by
+  simp only [gaussSum, AddChar.inv_apply, Finset.sum_mul, Finset.mul_sum, MulChar.inv_apply']
+  conv =>
+    lhs; congr; next => skip
+    ext; congr; next => skip
+    ext
+    rw [mul_mul_mul_comm, ← map_mul, ← map_add_mul, ← sub_eq_add_neg]
+--  conv in _ * _ * (_ * _) => rw [mul_mul_mul_comm, ← map_mul, ← map_add_mul, ← sub_eq_add_neg]
+  simp_rw [gaussSum_mul_aux hχ ψ]
+  rw [Finset.sum_comm]
+  classical -- to get `[DecidableEq R]` for `sum_mulShift`
+  simp_rw [← Finset.mul_sum, sum_mulShift _ hψ, sub_eq_zero, apply_ite, Nat.cast_zero, mul_zero]
+  rw [Finset.sum_ite_eq' Finset.univ (1 : R)]
+  simp only [Finset.mem_univ, map_one, one_mul, if_true]
+#align gauss_sum_mul_gauss_sum_eq_card gaussSum_mul_gaussSum_eq_card
+
+/-- When `χ` is a nontrivial quadratic character, then the square of `gaussSum χ ψ`
+is `χ(-1)` times the cardinality of `R`. -/
+theorem gaussSum_sq {χ : MulChar R R'} (hχ₁ : IsNontrivial χ) (hχ₂ : IsQuadratic χ)
+    {ψ : AddChar R R'} (hψ : IsPrimitive ψ) : gaussSum χ ψ ^ 2 = χ (-1) * Fintype.card R := by
+  rw [pow_two, ← gaussSum_mul_gaussSum_eq_card hχ₁ hψ, hχ₂.inv, mul_rotate']
+  congr
+  rw [mul_comm, ← gaussSum_mulShift _ _ (-1 : Rˣ), inv_mulShift]
+  rfl
+#align gauss_sum_sq gaussSum_sq
+
+end GaussSumProd
+
+/-!
+### Gauss sums and Frobenius
+-/
+
+
+section gaussSum_frob
+
+variable {R : Type u} [CommRing R] [Fintype R] {R' : Type v} [CommRing R']
+
+-- We assume that the target ring `R'` has prime characteristic `p`.
+variable (p : ℕ) [fp : Fact p.Prime] [hch : CharP R' p]
+
+/-- When `R'` has prime characteristic `p`, then the `p`th power of the Gauss sum
+of `χ` and `ψ` is the Gauss sum of `χ^p` and `ψ^p`. -/
+theorem gaussSum_frob (χ : MulChar R R') (ψ : AddChar R R') :
+    gaussSum χ ψ ^ p = gaussSum (χ ^ p) (ψ ^ p) := by
+  rw [← frobenius_def, gaussSum, gaussSum, map_sum]
+  simp_rw [pow_apply' χ fp.1.pos, map_mul, frobenius_def]
+  rfl
+#align gauss_sum_frob gaussSum_frob
+
+/-- For a quadratic character `χ` and when the characteristic `p` of the target ring
+is a unit in the source ring, the `p`th power of the Gauss sum of`χ` and `ψ` is
+`χ p` times the original Gauss sum. -/
+-- Porting note: TODO
```

Comment:
What is todo?

Reply:
Oh yes, I forgot to say that I added `norec`

Reply:
I guess there is no way to modify this PR once it's merged...